binding.c

```diff
@@ -11,24 +99,2109 @@ napi_value sn_randombytes_random (napi_env env, napi_callback_info info) {
+  SN_ARGV_UINT32(unpadded_buflen, 1)
+  SN_ARGV_UINT32(blocksize, 2)
+
+  SN_THROWS(unpadded_buflen > buf_size, "unpadded length cannot exceed buffer length")
+  SN_THROWS(blocksize > buf_size, "block size cannot exceed buffer length")
+
+  napi_value result;
+  size_t padded_buflen;
+  sodium_pad(&padded_buflen, buf_data, unpadded_buflen, blocksize, buf_size);
+  assert(napi_create_uint32(env, padded_buflen, &result) == napi_ok);
+  return result;
+}
+
+napi_value sn_sodium_unpad(napi_env env, napi_callback_info info) {
+  SN_ARGV(3, sodium_unpad);
+
+  SN_ARGV_TYPEDARRAY(buf, 0)
+  SN_ARGV_UINT32(padded_buflen, 1)
+  SN_ARGV_UINT32(blocksize, 2)
+
+  SN_THROWS(padded_buflen > buf_size, "unpadded length cannot exceed buffer length")
+  SN_THROWS(blocksize > buf_size, "block size cannot exceed buffer length")
+
+  napi_value result;
+  size_t unpadded_buflen;
+  sodium_unpad(&unpadded_buflen, buf_data, padded_buflen, blocksize);
+  assert(napi_create_uint32(env, unpadded_buflen, &result) == napi_ok);
+  return result;
+}
+
+napi_value sn_crypto_sign_keypair(napi_env env, napi_callback_info info) {
+  SN_ARGV(2, crypto_sign_keypair)
+
+  SN_ARGV_TYPEDARRAY(pk, 0)
+  SN_ARGV_TYPEDARRAY(sk, 1)
+
+  SN_THROWS(pk_size != crypto_sign_PUBLICKEYBYTES, "public key must be 32 bytes")
+  SN_THROWS(sk_size != crypto_sign_SECRETKEYBYTES, "secret key must be 64 bytes")
+
+  SN_RETURN(crypto_sign_keypair(pk_data, sk_data), "keypair generation failed")
+}
+
+napi_value sn_crypto_sign_seed_keypair(napi_env env, napi_callback_info info) {
+  SN_ARGV(3, crypto_sign_seed_keypair)
+
+  SN_ARGV_TYPEDARRAY(pk, 0)
+  SN_ARGV_TYPEDARRAY(sk, 1)
+  SN_ARGV_TYPEDARRAY(seed, 2)
+
+  SN_THROWS(pk_size != crypto_sign_PUBLICKEYBYTES, "public key must be 32 bytes")
+  SN_THROWS(sk_size != crypto_sign_SECRETKEYBYTES, "secret key must be 64 bytes")
+  SN_THROWS(seed_size != crypto_sign_SEEDBYTES, "seed must be 32 bytes")
+
+  SN_RETURN(crypto_sign_seed_keypair(pk_data, sk_data, seed_data), "keypair generation failed")
+}
+
+napi_value sn_crypto_sign(napi_env env, napi_callback_info info) {
+  SN_ARGV(3, crypto_sign)
+
+  SN_ARGV_TYPEDARRAY(signed_message, 0)
+  SN_ARGV_TYPEDARRAY(message, 1)
+  SN_ARGV_TYPEDARRAY(sk, 2)
+
+  SN_THROWS(signed_message_size != crypto_sign_BYTES + message_size, "signed message buffer must be 64 bytes longer than input")
+  SN_THROWS(sk_size != crypto_sign_SECRETKEYBYTES, "secret key must be 64 bytes")
+
+  SN_RETURN(crypto_sign(signed_message_data, NULL, message_data, message_size, sk_data), "signature failed")
+}
+
+napi_value sn_crypto_sign_open(napi_env env, napi_callback_info info) {
+  SN_ARGV(3, crypto_sign_open)
+
+  SN_ARGV_TYPEDARRAY(message, 0)
+  SN_ARGV_TYPEDARRAY(signed_message, 1)
+  SN_ARGV_TYPEDARRAY(pk, 2)
+
+  SN_THROWS(message_size != signed_message_size - crypto_sign_BYTES, "message buffer must be 64 bytes shorter than input")
+  SN_THROWS(signed_message_size < crypto_sign_BYTES, "signed message must be at least 64 bytes")
```

Comment:
should this not be != also?

Reply:
`sm` can be any length > `crypto_sign_BYTES` here since `m` is prepended